### PR TITLE
Strict tiling

### PR DESCRIPTION
**Pull request by balt-dev**
Tiling mode is now an enumerated string, out-of-bounds animation frames now give an error, and many tiles have been fixed.

### CODE REVIEW

File: README.md
Comment (CenTdemeern1) [inline README.md:1]: Is there a reason why this PR includes unrelated README changes?

File: config.py
Comment (CenTdemeern1) [inline config.py:1]: I'm not sure you meant to commit this

File: data/custom/baba.toml
Comment (CenTdemeern1) [inline data/custom/baba.toml:1]: This file seems to be using the old header comment. Are there any others like it?

File: data/levels/levels/world_data.txt
Comment (CenTdemeern1) [inline data/levels/levels/world_data.txt:1]: Was this file meant to be deleted?

File: data/palettes/abstract.png
Comment (CenTdemeern1) [inline data/palettes/abstract.png:1]: Why did so many unrelated sprites and such change? Is this metadata? Should we be using optipng?

File: data/sprites/bab/txt_''_0_1.png
Comment (CenTdemeern1) [inline data/sprites/bab/txt_''_0_1.png:1]: Wait, are '' and "" allowed?

File: data/sprites/bab/txt_uncolon)_0_1.png
Comment (CenTdemeern1) [inline data/sprites/bab/txt_uncolon)_0_1.png:1]: This file name seems like a really weird and awkward mix

File: src/types.py [hunk 1]
```diff
@@ -33,6 +33,9 @@ async def send(self,
     async def warn(self, msg: str, **kwargs) -> discord.Message: ...
 
 
+class Database:
+    ...
```
Comment (CenTdemeern1) [inline src/types.py:37]: Database stub? Ellipsis?

File: src/types.py [hunk 2]
```diff
@@ -54,10 +57,7 @@ def __init__(
             prefixes: list[str],
             exit_code: int = 0,
             **kwargs):
-        super().__init__(*args, **kwargs)
-        self.baba_loaded: bool = False
-        self.macros: dict = None
-        self.flags = None
+        ...
```
Comment (CenTdemeern1) [inline src/types.py:60]: Any reason why you're leaving this out?

File: src/types.py [hunk 3]
```diff
@@ -323,3 +323,59 @@ class BuiltinMacro:
...
+    def __str__(self) -> str:
+        if self == TilingMode.CUSTOM: return "custom"
+        if self == TilingMode.NONE: return "none"
+        if self == TilingMode.DIRECTIONAL: return "directional"
+        if self == TilingMode.TILING: return "tiling"
+        if self == TilingMode.CHARACTER: return "character"
+        if self == TilingMode.ANIMATED_DIRECTIONAL: return "animated_directional"
+        if self == TilingMode.ANIMATED: return "animated"
+        if self == TilingMode.STATIC_CHARACTER: return "static_character"
+        if self == TilingMode.DIAGONAL_TILING: return "diagonal_tiling"
```
Comment (CenTdemeern1) [inline src/types.py:348]: Still not too fussed by if chaining but you do you

File: src/db.py [hunk 4]
```diff
@@ -284,24 +286,38 @@ class TileData:
...
+                extra_frames = None
+
         return TileData(
             row["name"],
             row["sprite"],
             row["source"],
             (row["inactive_color_x"], row["inactive_color_y"]),
             (row["active_color_x"], row["active_color_y"]),
-            row["tiling"],
+            TilingMode(row["tiling"]),
```
Comment (CenTdemeern1) [inline src/db.py:316]: An invalid tiling mode here seems like undefined behavior

File: src/cogs/utilities.py [hunk 5]
```diff
@@ -153,7 +158,7 @@ async def search(self, ctx: Context, *, query: str = ""):
         * `source`: The source of the sprite. This should be a sprite mod.
         * `modded`: Whether to only return modded tiles (either `true` or `false`).
         * `color`: The color of the sprite. This can be a color name (`red`) or a palette (`0/3`).
-        * `tiling`: The tiling type of the object. This must be one of `-1`, `0`, `1`, `2`, `3` or `4`.
+        * `tiling`: The tiling type of the object.
```
Comment (CenTdemeern1) [inline src/cogs/utilities.py:161]: Should the different tiling mode options be listed here?

File: src/cogs/errorhandler.py [hunk 6]
```diff
@@ -188,8 +188,6 @@ async def on_command_error(self, ctx: Context, error: Exception):
 
             elif isinstance(error, numpy.linalg.LinAlgError):
                 return await ctx.error("The given warp points are unsolvable.")
-            elif isinstance(error, sqlite3.OperationalError):
-                return await ctx.error(f"There was an error in your SQL.\n`{error}`")
```
Comment (CenTdemeern1) [inline src/cogs/errorhandler.py:192]: This seems like it shouldn't be removed

File: src/cogs/owner.py [hunk 7]
```diff
@@ -321,10 +334,17 @@ async def loadbab(self, ctx: Context):
...
+                name.replace("<", "lt")
+                    .replace(">", "gt")
+                    .replace(":", "colon")
+                    .replace("\"", "quot")
+                    .replace("/", "sol")
+                    .replace("\\", "bsol")
+                    .replace("|", "vert")
+                    .replace("?", "quest")
+                    .replace("*", "ast")
+            )
```
Comment (CenTdemeern1) [inline src/cogs/owner.py:347]: This seems like it could become a problem (also, no handling of `'`?)

File: src/cogs/owner.py [hunk 8]
```diff
@@ -536,6 +556,10 @@ def prepare(d: dict[str, Any]) -> dict[str, Any]:
                 active_x = int(a_x)
                 active_y = int(a_y)
             tiling = int(tiling)
+            if tiling == TilingMode.TILING:
+                # Check for diagonal tiling
+                if pathlib.Path(f"data/sprites/{constants.BABA_WORLD}/{sprite}_16_1.png").exists():
+                    tiling = +TilingMode.DIAGONAL_TILING
```
Comment (CenTdemeern1) [inline src/cogs/owner.py:562]: Seems jank, but fine I guess

File: src/cogs/owner.py [hunk 9]
```diff
@@ -619,7 +643,8 @@ def freeze_dict(d: dict[str, Any]) -> tuple[tuple[str, Any], ...]:
                 :tiling,
                 :text_type,
                 NULL,
-                ""
+                '',
+                ''
```
Comment (CenTdemeern1) [inline src/cogs/owner.py:647]: What?

File: src/cogs/owner.py [hunk 10]
```diff
@@ -661,6 +686,10 @@ async def load_editor_tiles(self):
             inactive_x = int(c_x)
             inactive_y = int(c_y)
             tiling = int(tiling)
+            if tiling == TilingMode.TILING:
+                # Check for diagonal tiling
+                if pathlib.Path(f"data/sprites/{constants.BABA_WORLD}/{sprite}_16_1.png").exists():
+                    tiling = +TilingMode.DIAGONAL_TILING
```
Comment (CenTdemeern1) [inline src/cogs/owner.py:692]: Copy pasted check from earlier?

File: src/cogs/owner.py [hunk 11]
```diff
@@ -919,21 +956,7 @@ async def makedir(self, ctx: Context, name: str):
...
-# 
-#     Note: If your tile has tiling type 1,
-#           you must add the "diagonal = true/false" attribute to the tile
-#           in order for it to pass CI.
-#           For example, "line" = { ..., diagonal = false }. 
-# 
-# 
-# ----------------------------------------------------------------------------            
-""")
+            f.write("# Please read CONTRIBUTING.md for guidance on how to properly edit this file.\n\n\n")
```
Comment (CenTdemeern1) [inline src/cogs/owner.py:959]: Also seems copy pasted?

**Review by CenTdemeern1 — CHANGES_REQUESTED** [review]
A few concerns and mostly nitpicks (this many file changes weren't fun to review... my browser is lagging...)

File: data/sprites/bab/txt_uncolon)_0_1.png
Comment (balt-dev) [reply data/sprites/bab/txt_uncolon)_0_1.png:1]: i added a manual tilename override specifically because of that

File: data/sprites/bab/txt_''_0_1.png
Comment (balt-dev) [reply data/sprites/bab/txt_''_0_1.png:1]: in filenames in windows and unix, surprisingly yes

File: data/palettes/abstract.png
Comment (balt-dev) [reply data/palettes/abstract.png:1]: i uh accidentally committed an old version of the basegame palettes so i fixed that here. should've done it on main but i forgot to switch branches

File: data/levels/levels/world_data.txt
Comment (balt-dev) [reply data/levels/levels/world_data.txt:1]: no but it serves zero purpose so it's probably fine

File: data/custom/baba.toml
Comment (balt-dev) [reply data/custom/baba.toml:1]: not  that i'm aware of?

File: config.py
Comment (balt-dev) [reply config.py:1]: no i did not! good catch thanks

File: src/types.py (hunk 1 above)
Comment (balt-dev) [reply src/types.py:37]: yeah, it's a stub

File: src/types.py (hunk 2 above)
Comment (balt-dev) [reply src/types.py:60]: src/types is for stubs, i probably should move TilingMode back to constants

File: src/db.py (hunk 4 above)
Comment (balt-dev) [reply src/db.py:316]: as long as what's in the database is fine then this is fine, but it probably would be better to add checking for it

File: src/cogs/utilities.py (hunk 5 above)
Comment (balt-dev) [reply src/cogs/utilities.py:161]: they should but i got lazy

File: src/cogs/errorhandler.py (hunk 6 above)
Comment (balt-dev) [reply src/cogs/errorhandler.py:192]: incorrect sql mostly comes from internal errors, so it's more useful to get a traceback

File: src/cogs/owner.py (hunk 9 above)
Comment (balt-dev) [reply src/cogs/owner.py:647]: sqlite is supposed to use single quotes for string literals, and a row was added for extra_frames

File: src/cogs/owner.py (hunk 10 above)
Comment (balt-dev) [reply src/cogs/owner.py:692]: i think so

File: src/cogs/owner.py (hunk 11 above)
Comment (balt-dev) [reply src/cogs/owner.py:959]: think so here too

File: src/types.py (hunk 3 above)
Comment (balt-dev) [reply src/types.py:348]: match is about the same level of performance (actually marginally worse) and adds an extra indentation level

File: README.md
Comment (balt-dev) [reply README.md:1]: i forgot to switch branches

File: data/custom/baba.toml
Comment (balt-dev) [reply data/custom/baba.toml:1]: come to think of it this should probably be reimported because i imported it using an older version of the game, but i'd need a copy of the latest version of baba is you for that which i don't. have at the moment, because i'm at college

File: README.md
Comment (CenTdemeern1) [reply README.md:1]: Please just cherrypick your commit to the right branch or something, if I really have to go through 249 files by hand I'd refer you'd only keep it to the relevant ones

File: config.py
Comment (CenTdemeern1) [reply config.py:1]: (I'll mark this conversation as resolved when you fix this)

File: data/custom/baba.toml
Comment (CenTdemeern1) [reply data/custom/baba.toml:1]: I do. If you'd like, and if you can give me proper instructions, I can do it for you.

File: data/palettes/abstract.png
Comment (CenTdemeern1) [reply data/palettes/abstract.png:1]: See convo 1

File: data/sprites/bab/txt_''_0_1.png
Comment (CenTdemeern1) [reply data/sprites/bab/txt_''_0_1.png:1]: Sure I guess? But this kind of goes against the whole "denylisting special characters" thing we discussed

File: data/sprites/bab/txt_uncolon)_0_1.png
Comment (CenTdemeern1) [reply data/sprites/bab/txt_uncolon)_0_1.png:1]: See previous conversation

File: src/types.py (hunk 1 above)
Comment (CenTdemeern1) [reply src/types.py:37]: Why?

File: src/types.py (hunk 2 above)
Comment (CenTdemeern1) [reply src/types.py:60]: Yeah I was wondering why you moved it. Maybe move it back then

File: src/types.py (hunk 3 above)
Comment (CenTdemeern1) [reply src/types.py:348]: This is a hard violation of DRY though, to an unreasonable degree
I know DRY isn't enforced here but this is getting out of hand

File: src/db.py (hunk 4 above)
Comment (CenTdemeern1) [reply src/db.py:316]: I strongly suggest adding checking here.

File: src/cogs/utilities.py (hunk 5 above)
Comment (CenTdemeern1) [reply src/cogs/utilities.py:161]: I can tell. Please fix this.

File: src/cogs/errorhandler.py (hunk 6 above)
Comment (CenTdemeern1) [reply src/cogs/errorhandler.py:192]: Sure.

File: src/cogs/owner.py (hunk 9 above)
Comment (CenTdemeern1) [reply src/cogs/owner.py:647]: Ah.

File: src/cogs/owner.py (hunk 10 above)
Comment (CenTdemeern1) [reply src/cogs/owner.py:692]: Is there anything that reasonably can or should be done about this?

File: src/cogs/owner.py (hunk 11 above)
Comment (CenTdemeern1) [reply src/cogs/owner.py:959]: This might need to be a constant or something.

**Comment by CenTdemeern1** [comment]
Most of this boils down to "please clean up this branch and its diff"

**Comment by CenTdemeern1** [comment]
@balt-dev your turn.

File: src/cogs/owner.py (hunk 7 above)
Comment (balt-dev) [reply src/cogs/owner.py:347]: what do you think i should do?

**Comment by balt-dev** [comment]
give me a little while, i'm in class

**Comment by balt-dev** [comment]
thanks for doing this btw

File: src/cogs/owner.py (hunk 7 above)
Comment (CenTdemeern1) [reply src/cogs/owner.py:347]: I'm not sure, it's just that this doesn't catch all special characters and endlessly replace chaining seems like a bad idea 
I'd prefer something that just applies to certain character ranges but I'm not sure where the character names would come from

**Comment by balt-dev** [comment]
i'll clean this up a bit and merge it, it's been a blocker for a while on new tiles

File: src/cogs/owner.py (hunk 11 above)
Comment (balt-dev) [reply src/cogs/owner.py:959]: i searched around and it seems this is the only occurrence of this string now so idk

File: src/cogs/owner.py (hunk 10 above)
Comment (balt-dev) [reply src/cogs/owner.py:692]: in all honestly this is likely fine to just leave as is

File: src/cogs/owner.py (hunk 7 above)
Comment (balt-dev) [reply src/cogs/owner.py]: Resolved by latest commit

File: config.py
Comment (balt-dev) [reply config.py:1]: Resolved by latest commit

File: src/cogs/owner.py (hunk 10 above)
Comment (balt-dev) [reply src/cogs/owner.py:687]: this was copied from convert_json.py which no longer exists so :P

**Comment by balt-dev** [comment]
yes this is a lot of non-pr-related changes
no i don't care working on this codebasde is a fucking nigthmare please just let me merge this

**Comment by balt-dev** [comment]
charlotte i'm so sorry but i cannot handle review right now

**Comment by CenTdemeern1** [comment]
OkayI know you’re frustrated, please calm downI don’t care about the review I’d prefer if you just took a break for your mental wellbeing right now. I doubt people would mind that much if you need it this badlyOn 22 Oct 2024, at 1:34 AM, Balt ***@***.***> wrote:﻿
charlotte i'm so sorry but i cannot handle review right now
